memq-client/src/main/java/com/pinterest/memq/commons/storage/s3express/SessionTokenManager.java

```diff
@@ -0,0 +1,162 @@
+  private Map<String, ConcurrentLinkedDeque<SessionCreds>> bucketCredentialMap = new ConcurrentHashMap<>();
+  private ScheduledExecutorService es = Executors.newScheduledThreadPool(1, new ThreadFactory() {
+
+    @Override
+    public Thread newThread(Runnable r) {
+      Thread th = new Thread(r);
+      th.setDaemon(true);
+      return th;
+    }
+  });
+  private HttpClient secureClient;
+
+  public SessionTokenManager() {
+    ConnectionProvider provider = ConnectionProvider.builder("s3express").maxConnections(10)
+        .maxIdleTime(Duration.ofSeconds(20)).maxLifeTime(Duration.ofSeconds(60))
+        .pendingAcquireTimeout(Duration.ofSeconds(60)).evictInBackground(Duration.ofSeconds(120))
+        .build();
+    secureClient = HttpClient.create(provider).option(ChannelOption.SO_SNDBUF, 4 * 1024 * 1024)
+        .option(ChannelOption.SO_LINGER, 0).secure();
+  }
+
+  public static SessionTokenManager getInstance() {
+    return mgr;
+  }
+
+  public SessionCreds getCredentials(final String bucketname) throws InterruptedException {
+    ConcurrentLinkedDeque<SessionCreds> concurrentLinkedDeque = bucketCredentialMap.get(bucketname);
+    if (concurrentLinkedDeque == null) {
+      synchronized (bucketCredentialMap) {
+        concurrentLinkedDeque = bucketCredentialMap.get(bucketname);
+        if (concurrentLinkedDeque == null) {
+          concurrentLinkedDeque = new ConcurrentLinkedDeque<>();
+          // start the scheduled task for credential refresh
+          final ConcurrentLinkedDeque<SessionCreds> concurrentLinkedDequeRef = concurrentLinkedDeque;
+          es.scheduleAtFixedRate(() -> {
+            try {
+              SessionCreds fetchCredentials = fetchCredentials(bucketname);
+              concurrentLinkedDequeRef.add(fetchCredentials);
+              if (concurrentLinkedDequeRef.size() == 2) {
+                // purge existing credentials
+                concurrentLinkedDequeRef.poll();
+              }
+            } catch (Exception e) {
+              e.printStackTrace();
+            }
+          }, 0, 4, TimeUnit.MINUTES);
+          bucketCredentialMap.put(bucketname, concurrentLinkedDeque);
+        }
+      }
+    }
+    while (concurrentLinkedDeque.isEmpty()) {
+      Thread.sleep(100);
+    }
+    return concurrentLinkedDeque.peek();
+  }
+
+  protected SessionCreds fetchCredentials(String bucketname) throws Exception {
+    InstanceProfileCredentialsProvider credentialProvider = InstanceProfileCredentialsProvider
+        .builder().asyncCredentialUpdateEnabled(true).asyncThreadName("IamCredentialUpdater")
+        .build();
+    // Sign it...
+    AwsS3V4Signer signer = AwsS3V4Signer.create();
+    SdkHttpFullRequest req = SdkHttpFullRequest.builder()
+        .appendHeader("x-amz-create-session-mode", "ReadWrite")
+        .appendRawQueryParameter("session", "").method(SdkHttpMethod.GET)
+        .uri(URI.create("https://" + bucketname + ".s3express-use1-az5.us-east-1.amazonaws.com"))
```

Comment:
Region and AZ are hardcoded in the URL template. I think these two values should come from method inputs.

Reply:
Yes, but we need to use the lookup table that needs to be added here.